### PR TITLE
[TWEAK] Runechat Color Preference and Species Dependent Organ (runechat) Removal

### PR DESCRIPTION
**Pull request by Spaghetti-bit**
## What Does This PR Do

Adds a new preference for users to change their runechat color character by character without needing to rely on their hair color, eye color, or skin color.

**This removes the ability for specific races to have their runechat color be dictated via their species specific color changes.

Removes the preference toggle to force white runechat.

Summarized / TLDR
- changes the SQL query to include the new preferences.
- allows people to change their runechat color via mirror or character creator.
- Default runechat color is #FFFFFF => White.

## Why It's Good For The Game

Having to change your character's eyes, hair, etc to get a runechat color that is unique shouldn't be a thing, this allows people to customize their character just that tiny bit more

## Images of changes

image

image
image

## Testing

image

Tested changeling transform, polymorph mutation, creating new characters, loading slots, gibbing and reincarnating, changing eye color via mirror. Lots of testing.

Tested to make sure the DNA cannot be modified via radiation or DNA modifier.

## Changelog
:cl:
add: Runechat color preference.
/:cl:

### CODE REVIEW

File: code/modules/client/preference/character.dm [hunk 1]
```diff
@@ -476,11 +481,12 @@
...
-	physique = query.item[56]
-	height = query.item[57]
-	cyborg_brain_type = query.item[58]
-	body_type = query.item[59]
+	runechat_color = query.item[55]
+	var/custom_emotes_tmp = query.item[56]
+	physique = query.item[57]
+	height = query.item[58]
+	cyborg_brain_type = query.item[59]
+	body_type = query.item[60]
```
Comment (AffectedArc07) [inline code/modules/client/preference/character.dm:489]: You need to also edit `20-load_characters.dm` to actually have it read from this query.

Please test this code with a database properly.

File: SQL/paradise_schema.sql [hunk 2]
```diff
@@ -79,6 +79,7 @@ CREATE TABLE `characters` (
   `hair_gradient_offset` varchar(7) COLLATE utf8mb4_unicode_ci NOT NULL DEFAULT '0,0',
   `hair_gradient_colour` varchar(7) COLLATE utf8mb4_unicode_ci NOT NULL DEFAULT '#000000',
   `hair_gradient_alpha` tinyint(3) UNSIGNED NOT NULL DEFAULT '255',
+  `runechat_color` varchar(7) COLLATE utf8mb4_unicode_ci NOT NULL DEFAULT '#000000',
```
Comment (AffectedArc07) [inline SQL/paradise_schema.sql:82]: Default this to something other than black *please* otherwise no one will have readable runechat. Make it `#fffff` at a minimum.

**Review by AffectedArc07 — CHANGES_REQUESTED** [review]
You're missing the SQL 57-58 update file.

File: code/modules/client/preference/character.dm (hunk 1 above)
Comment (Spaghetti-bit) [reply code/modules/client/preference/character.dm:489]: image

Tested properly, pushing a commit.

File: SQL/paradise_schema.sql (hunk 2 above)
Comment (Spaghetti-bit) [reply SQL/paradise_schema.sql]: Defaulted to #FFFFFF, white for readability.

File: code/modules/client/preference/character.dm (hunk 1 above)
Comment (Spaghetti-bit) [reply code/modules/client/preference/character.dm:489]: `code/__DEFINES/misc_defines.dm` and the configuration `toml` updated.

Comment (Spaghetti-bit) [reply code/modules/client/preference/character.dm:489]: Here is the query with some characters with differing runechat colors @AffectedArc07 
image

Comment (Spaghetti-bit) [reply code/modules/client/preference/character.dm:489]: image

File: code/modules/tgui/modules/appearance_changer.dm [hunk 3]
```diff
@@ -141,6 +141,11 @@
 				if(new_eyes && (!..()) && owner.change_eye_color(new_eyes))
 					update_dna()
 
+		if("runechat_color")
+			var/new_runechat_color = input("Please select runechat color.", "Runechat Color", owner.runechat_color) as color|null
+			if(new_runechat_color && (!..()) && owner.change_runechat_color(new_runechat_color))
+				update_dna()
```
Comment (Contrabang) [inline code/modules/tgui/modules/appearance_changer.dm:147]: update_dna is part of your new proc
```suggestion
			if(new_runechat_color && (!..()))
				owner.change_runechat_color(new_runechat_color)
```

File: code/modules/mob/living/carbon/human/species/_species.dm [hunk 4]
```diff
@@ -153,6 +153,9 @@
 	var/default_headacc				//Default head accessory style for newly created humans unless otherwise set.
 	var/default_headacc_colour
 
+	///Runechat color, stored on the mob itself and not in preferances.
+	var/default_runechat_color = "#FFFFFF"
```
Comment (Contrabang) [inline code/modules/mob/living/carbon/human/species/_species.dm:157]: This isn't ever used
```suggestion
```

File: code/modules/mob/living/carbon/human/appearance.dm [hunk 5]
```diff
@@ -235,6 +235,14 @@
 		update_body()
 	return TRUE
 
+/mob/living/carbon/human/proc/change_runechat_color(colour = "#000000", update_dna = TRUE, skip_icons = TRUE)
```
Comment (Contrabang) [inline code/modules/mob/living/carbon/human/appearance.dm:238]:
```suggestion
/mob/living/carbon/human/proc/change_runechat_color(colour = "#000000", update_dna = TRUE)
```

Comment (Contrabang) [inline code/modules/mob/living/carbon/human/appearance.dm:238]: when would you ever want `update_dna ` to be false? I dont think theres any proc instances here where its false

**Review by Contrabang — DISMISSED** [review]

File: code/modules/mob/living/carbon/human/appearance.dm (hunk 5 above)
Comment (Spaghetti-bit) [reply code/modules/mob/living/carbon/human/appearance.dm]: Never, this was an oversight from me taking reference of the proc above it, `change_eye_color`

File: code/modules/mob/living/carbon/human/species/_species.dm [hunk 6]
```diff
@@ -1053,17 +1054,12 @@ It'll return null if the organ doesn't correspond, so include null checks when u
...
   * Arguments:
   * * H - The human who this DNA belongs to
   */
 /datum/species/proc/get_species_runechat_color(mob/living/carbon/human/H)
-	if(bodyflags & HAS_SKIN_COLOR)
-		return H.skin_colour
-	else
-		var/obj/item/organ/external/head/HD = H.get_organ("head")
-		return HD?.hair_colour
+	return H.runechat_color
```
Comment (AffectedArc07) [inline code/modules/mob/living/carbon/human/species/_species.dm:1062]: This proc is useless, make the caller just refer to this var

Comment (Spaghetti-bit) [reply code/modules/mob/living/carbon/human/species/_species.dm:1062]: this was a left over i forgot to remove.

Comment (Spaghetti-bit) [reply code/modules/mob/living/carbon/human/species/_species.dm]: Completely axed, resolving as it is outdated.

File: code/modules/client/preference/character.dm [hunk 7]
```diff
@@ -1910,6 +1917,9 @@
 	character.change_eye_color(e_colour, skip_icons = TRUE)
 	character.original_eye_color = e_colour
 
+	//Runechat Color
```
Comment (1080pCat) [inline code/modules/client/preference/character.dm:1920]:
```suggestion
	// Runechat Color
```

File: code/modules/mob/living/carbon/human/human_defines.dm [hunk 8]
```diff
@@ -96,6 +96,9 @@
 	var/list/splinted_limbs = list() //limbs we know are splinted
 	var/original_eye_color = "#000000"
 
+	///Runechat color
```
Comment (1080pCat) [inline code/modules/mob/living/carbon/human/human_defines.dm:99]:
```suggestion
	/// Runechat color
```

File: code/modules/mob/living/carbon/human/species/_species.dm [hunk 9]
```diff
@@ -153,6 +153,7 @@
 	var/default_headacc				//Default head accessory style for newly created humans unless otherwise set.
 	var/default_headacc_colour
 
+
```
Comment (1080pCat) [inline code/modules/mob/living/carbon/human/species/_species.dm:156]:
```suggestion
```

**Review by 1080pCat — COMMENTED** [review]
shush, stale bot.
Also will this auto take existing chars hair colour to be the default? or will all text default to white

File: code/__DEFINES/misc_defines.dm [hunk 10]
```diff
@@ -405,7 +405,7 @@
 #define INVESTIGATE_HOTMIC "hotmic"
 
 // The SQL version required by this version of the code
-#define SQL_VERSION 57
+#define SQL_VERSION 58
```
Comment (Burzah) [inline code/__DEFINES/misc_defines.dm:408]:
```suggestion
#define SQL_VERSION 59
```

File: config/example/config.toml [hunk 11]
```diff
@@ -175,7 +175,7 @@ ipc_screens = [
 # Enable/disable the database on a whole
 sql_enabled = false
 # SQL version. If this is a mismatch, round start will be delayed
-sql_version = 57
+sql_version = 58
```
Comment (Burzah) [inline config/example/config.toml:178]:
```suggestion
sql_version = 59
```

File: SQL/paradise_schema.sql [hunk 12]
```diff
@@ -79,6 +79,7 @@ CREATE TABLE `characters` (
   `hair_gradient_offset` varchar(7) COLLATE utf8mb4_unicode_ci NOT NULL DEFAULT '0,0',
   `hair_gradient_colour` varchar(7) COLLATE utf8mb4_unicode_ci NOT NULL DEFAULT '#000000',
   `hair_gradient_alpha` tinyint(3) UNSIGNED NOT NULL DEFAULT '255',
+  `runechat_color` varchar(7) COLLATE utf8mb4_unicode_ci NOT NULL DEFAULT '#FFFFFF',
```
Comment (Burzah) [inline SQL/paradise_schema.sql:82]:
```suggestion
  `runechat_color` VARCHAR(7) COLLATE utf8mb4_unicode_ci NOT NULL DEFAULT '#FFFFFF',
```
Nit.

File: SQL/updates/57-58.sql [hunk 13]
```diff
@@ -0,0 +1,6 @@
+# Updates the DB from 57 to 58 ~SpaghettiBit
+# Adds a runechat color preference to characters.
+
+# Add runechat color after hair_gradient_alpha
+ALTER TABLE `characters`
+	ADD COLUMN `runechat_color` varchar(7) COLLATE utf8mb4_unicode_ci NOT NULL DEFAULT '#FFFFFF' AFTER `hair_gradient_alpha`;
```
Comment (Burzah) [inline SQL/updates/57-58.sql:6]:
```suggestion
# Updates the DB from 58 to 59 ~SpaghettiBit
# Adds a runechat color preference to characters.

# Add runechat color after hair_gradient_alpha
ALTER TABLE `characters`
	ADD COLUMN `runechat_color` VARCHAR(7) COLLATE utf8mb4_unicode_ci NOT NULL DEFAULT '#FFFFFF' AFTER `hair_gradient_alpha`;
```
This needs to be changed to 58-59

**Review by Burzah — CHANGES_REQUESTED** [review]

**Comment by 1080pCat** [comment]
I'll mirror some concerns i have here.
The default colour should be the hair(or species dependent organ). Not white.
Blanket changing *everyones* character setup to suddenly remove the rune chat text colour will cause quite alot of confusion.
Im not against having the ability to change the rune chat colour, its just retroactively making already existing character slots all white just feels wrong.

TLDR;
letting them change it is good.
blanket changing it to white for everyone, even retroactively is bad.

File: code/modules/client/preference/link_processing.dm [hunk 14]
```diff
@@ -391,6 +391,11 @@
 					if(result)
 						active_character.h_grad_colour = result
 
+				if("runechat_color")
+					var/result = input(user, "Choose your character's runechat color:", "Character Preference", active_character.runechat_color) as color|null
```
Comment (Burzah) [inline code/modules/client/preference/link_processing.dm:395]: Can you change this to use a TGUI input instead?

Comment (Burzah) [reply code/modules/client/preference/link_processing.dm:395]: Talked about in Discord. (Color wheel soon)™

**Review by Burzah — APPROVED** [review]

File: code/modules/client/preference/preferences.dm [hunk 15]
```diff
@@ -206,6 +206,8 @@ GLOBAL_LIST_INIT(special_role_times, list( //minimum age (in days) for accounts
 			dat += "<b>Gender:</b> <a href='byond://?_src_=prefs;preference=gender'>[active_character.gender == MALE ? "Male" : (active_character.gender == FEMALE ? "Female" : "Genderless")]</a><br>"
 			dat += "<b>Body Type:</b> <a href='byond://?_src_=prefs;preference=body_type'>[active_character.body_type == MALE ? "Masculine" : "Feminine"]</a>"
 			dat += "<br>"
+			dat += "<b>Runechat Color:</b> "
+			dat += "<a href='byond://?_src_=prefs;preference=runechat_color;task=input'>Color</a> [color_square(active_character.runechat_color)]<br>"
```
Comment (AffectedArc07) [inline code/modules/client/preference/preferences.dm:210]: `task=input` is redundant here I believe

File: code/modules/mob/living/carbon/human/human_mob.dm [hunk 16]
```diff
@@ -1931,7 +1931,7 @@ Eyes need to have significantly high darksight to shine unless the mob has the X
 /mob/living/carbon/human/get_runechat_color()
 	if(client?.prefs.toggles2 & PREFTOGGLE_2_FORCE_WHITE_RUNECHAT)
 		return "#FFFFFF" // Force white if they want it
-	return dna.species.get_species_runechat_color(src)
+	return runechat_color
```
Comment (AffectedArc07) [inline code/modules/mob/living/carbon/human/human_mob.dm:1934]: The above toggle is also kinda redundant now that its just a blanket colour option

File: tgui/packages/tgui/interfaces/AppearanceChanger.js [hunk 17]
```diff
@@ -54,6 +55,7 @@ export const AppearanceChanger = (props, context) => {
     change_eye_color ||
     change_skin_tone ||
     change_skin_color ||
+    change_runechat_color ||
```
Comment (AffectedArc07) [inline tgui/packages/tgui/interfaces/AppearanceChanger.js:58]: Can we make this `true` instead of passing `TRUE` from the TGUI data send each time?

**Review by AffectedArc07 — COMMENTED** [review]
Review done - apologies for fragmenting it.

Most of these are just queries as opposed to major changes

File: code/modules/mob/living/carbon/human/human_mob.dm (hunk 16 above)
Comment (Spaghetti-bit) [reply code/modules/mob/living/carbon/human/human_mob.dm:1934]: We still have it in the case that they don't want their runechat to be determined by a specific color, if we were to remove it we'd have to update the SQL among other things. This also just allows people to toggle it at a whim, which I am all for.

File: code/modules/mob/living/carbon/human/appearance.dm [hunk 18]
```diff
@@ -235,6 +235,13 @@
 		update_body()
 	return TRUE
 
+/mob/living/carbon/human/proc/change_runechat_color(colour = "#000000")
+	if(colour == runechat_color)
+		return
+	runechat_color = colour
+	update_dna()
+	return TRUE
```
[the remark this answers is not in the document]
Comment (Spaghetti-bit) [reply code/modules/mob/living/carbon/human/appearance.dm:243]: Following the above proc for reference.

image

Will remove it for both.

File: tgui/packages/tgui/interfaces/AppearanceChanger.js (hunk 17 above)
Comment (Spaghetti-bit) [reply tgui/packages/tgui/interfaces/AppearanceChanger.js:58]: To keep in line with the rest of the appearance changes, I'll keep it as is. It makes it more readable and digestible.
The most closely related one would be `change_eye_color` which follows the exact same principle.

image
image

File: code/modules/client/preference/preferences.dm (hunk 15 above)
Comment (Spaghetti-bit) [reply code/modules/client/preference/preferences.dm:210]: It is following the same pattern as the rest of preferences.

File: code/modules/mob/living/carbon/human/human_mob.dm (hunk 16 above)
Comment (AffectedArc07) [reply code/modules/mob/living/carbon/human/human_mob.dm:1934]: It wouldnt need an SQL update if its just a pref toggle were not needing.

The force white toggle isn't needed if they can just manually set their colour to `#ffffff` in settings.

Comment (Spaghetti-bit) [reply code/modules/mob/living/carbon/human/human_mob.dm:1934]: That is fair, I'll remove it since it's no longer needed.

Comment (Spaghetti-bit) [reply code/modules/mob/living/carbon/human/human_mob.dm]: > It wouldnt need an SQL update if its just a pref toggle were not needing.
> 
> The force white toggle isn't needed if they can just manually set their colour to `#ffffff` in settings.

Alright, fully done. `PREFTOGGLE_2_FORCE_WHITE_RUNECHAT` is no more.

File: config/example/config.toml [hunk 19]
```diff
@@ -186,7 +186,7 @@ sql_database = "paradise_gamedb"
 # SQL server username
 sql_username = "root"
 # SQL server password
-sql_password = "please use something secure in a production environment"
+sql_password = "root"
```
Comment (Burzah) [inline config/example/config.toml:189]: Not sure if you meant to commit this?

Comment (Spaghetti-bit) [reply config/example/config.toml:189]: I did not.

Comment (AffectedArc07) [reply config/example/config.toml]: The classic `root/root`

**Review by Burzah — APPROVED** [review]
Looked again after reading all the comments and changes made after AA's review. I believe this is ready to go.

**Review by AffectedArc07 — APPROVED** [review]
